src/rdkafka_cgrp.c

```diff
@@ -3124,8 +4141,302 @@ rd_kafka_cgrp_max_poll_interval_check_tmr_cb (rd_kafka_timers_t *rkts,
+                                                        par->partition);
+                                continue;
+                        }
+                }
+        }
+
+        if (revoking->cnt == 0) {
+                rd_kafka_topic_partition_list_destroy(revoking);
+                revoking = NULL;
+        }
+
+        return revoking;
+}
+
+
+/**
+ * @brief Handle a new subscription that is modifying an existing subscription
+ *        in the COOPERATIVE case.
+ */
+static rd_kafka_resp_err_t
+rd_kafka_cgrp_modify_subscription (rd_kafka_cgrp_t *rkcg,
+                                   rd_kafka_topic_partition_list_t
+                                   *rktparlist) {
+        rd_kafka_topic_partition_list_t *unsubscribing_topics;
+        rd_kafka_topic_partition_list_t *revoking;
+        rd_list_t *tinfos;
+        rd_kafka_topic_partition_list_t *errored;
+        int metadata_age;
+
+        rkcg->rkcg_flags &= ~RD_KAFKA_CGRP_F_WILDCARD_SUBSCRIPTION;
+
+        if (rd_kafka_topic_partition_list_regex_cnt(rktparlist) > 0)
+                rkcg->rkcg_flags |= RD_KAFKA_CGRP_F_WILDCARD_SUBSCRIPTION;
+
+        /* Topics in rkcg_subscribed_topics that don't match any pattern in
+           the new subscription. */
+        unsubscribing_topics = rd_kafka_cgrp_calculate_unsubscribing_topics(
+                                        rkcg, rktparlist);
+
+        /* Currently assigned topic partitions to that are no longer
+           desired. */
+        revoking = rd_kafka_cgrp_calculate_subscribe_revoking_partitions(
+                                        rkcg, unsubscribing_topics);
+
+        if (unsubscribing_topics)
+                rd_kafka_topic_partition_list_destroy(unsubscribing_topics);
+
+        rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "SUBSCRIBE",
+                     "Group \"%.*s\": modifying subscription of size %d to "
+                     "new subscription of size %d%s (join state %s)",
+                     RD_KAFKAP_STR_PR(rkcg->rkcg_group_id),
+                     rkcg->rkcg_subscription->cnt, rktparlist->cnt,
+                     !unsubscribing_topics ? " without removing any topics" :
+                     (!revoking ? " without revoking any partitions" : ""),
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
+
+        rd_kafka_topic_partition_list_destroy(rkcg->rkcg_subscription);
+        rkcg->rkcg_subscription = rktparlist;
+
+        if (rd_kafka_cgrp_metadata_refresh(rkcg, &metadata_age,
+                                           "modify subscription") == 1) {
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "JOIN",
+                             "Group \"%.*s\": postponing join until "
+                             "up-to-date metadata is available",
+                             RD_KAFKAP_STR_PR(rkcg->rkcg_group_id));
+
+                if (revoking) {
+                        rd_kafka_rebalance_op_incr_maybe(
```

Comment:
this is an example of something that could be so much easier. with Tasks async/await. you'd have a metadata cache abstraction which you say `await give_me_metadata(force_update_maybe)` which encapsulates everything including possible concurrent calls to it. then on the next line, you use the metadata. everything related to scheduling, where it's run, worrying about the state machine transitions, finding the async logic scattered all over the place is not your concern. in librdkafka, all of these concerns are inter-mingled, and the complexity compounds.

anyway, where were we. yes, you are right. I think though that this is an edge case that won't come up very often at all, and what is here will work in all cases (just more rebalances than desirable). I think it's a case of easy winning over optimal.

back to the rant - in the higher level architecture specified above, optimal would be trivial.

Reply:
i'll add a comment pointing out your point for now. if you want to argue the point, we should discuss outside github.

Reply:
oh, right, haha, this can just be removed - the required rebalance is already implemented on metadata refresh.

my rant about this being difficult to follow, and that there are alternatives that are not still holds.